### PR TITLE
Cpp setup

### PR DESCRIPTION
**Pull request by Zolkin1**
Adding support for C++ code.

- clang-tidy and clang-format formatting
- C++ file structure and starter cmake files
- Place holder tests and library for the tests
- C++ dependencies in pixi
- C++ tests with Catch2

Additionally:

- Docker fixes for the dev contrainer

### CODE REVIEW

File: obelisk/cpp/lib1/include/place_holder.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Obelisk Software Framework.
+ * Developed in the Amber Lab at Caltech.
```
Comment (alberthli) [inline obelisk/cpp/lib1/include/place_holder.h:3]: nit: everywhere possible, fully capitalize AMBER

File: pixi.toml [hunk 2]
```diff
@@ -156,5 +162,22 @@ docs = { features=["docs"] }
 [tasks]
 # builds in the obelisk_ws directory only if src has changes
 source_rosbase = { cmd="bash source_rosbase.sh", cwd="scripts", env={ NOOB="true" } }
-build = { cmd="colcon build --symlink-install --parallel-workers $(nproc)", cwd="obelisk_ws", inputs=["src"], depends-on=["source_rosbase"] }
+ros-build = { cmd="colcon build --symlink-install --parallel-workers $(nproc)", cwd="obelisk_ws", inputs=["src"], depends-on=["source_rosbase"] }
```
Comment (alberthli) [inline pixi.toml:165]: we need to pick a consistent standard for tasks: either underscores or hyphens. i personally prefer underscores here, but you can make a case the other way.

File: pixi.toml [hunk 3]
```diff
@@ -126,6 +126,12 @@ ruff = ">=0.4.9"
 pyright = ">=1.1.368"
 
 
+# build
+[feature.dev.dependencies]
+cmake = "3.16.*"
+ninja = "1.12.1"
+make = ">=4.3,<5"
```
Comment (alberthli) [inline pixi.toml:133]: we should probably make a separate environment set for things that are required for building but not necessarily for development (e.g., if people do want to use pixi for running/building but don't want things like linters)

File: pixi.toml [hunk 4]
```diff
@@ -156,5 +162,22 @@ docs = { features=["docs"] }
 [tasks]
 # builds in the obelisk_ws directory only if src has changes
 source_rosbase = { cmd="bash source_rosbase.sh", cwd="scripts", env={ NOOB="true" } }
-build = { cmd="colcon build --symlink-install --parallel-workers $(nproc)", cwd="obelisk_ws", inputs=["src"], depends-on=["source_rosbase"] }
+ros-build = { cmd="colcon build --symlink-install --parallel-workers $(nproc)", cwd="obelisk_ws", inputs=["src"], depends-on=["source_rosbase"] }
 source_obelisk = { cmd="bash source_obelisk.sh", cwd="scripts", depends-on=["build"], env={ NOOB="true" } }
+
+# run the dummy tests
+cpp-test-dummy = {cmd = "obelisk/cpp/build/tests/dummy_test", depends-on = ["cpp-build"]}
```
Comment (alberthli) [inline pixi.toml:169]: same comment about hyphens/underscores

File: pixi.toml [hunk 5]
```diff
@@ -156,5 +162,22 @@ docs = { features=["docs"] }
...
+
+# run ctest
+cpp-ctest = {cmd = "ctest", cwd = "obelisk/cpp/build/tests", depends-on = "cpp-build"}
```
Comment (alberthli) [inline pixi.toml:172]: we might want a command that runs all tests, not just the c++ ones

File: pixi.toml (hunk 4 above)
Comment (alberthli) [reply pixi.toml:169]: also, i'm trying to be consistent about leaving spaces in curly braces:
```
{ cmd = "obelisk/cpp/build/tests/dummy_test", depends-on = ["cpp-build"] }
```

File: tests/tests-cpp/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,25 @@
+message(STATUS "Building obelisk_cpp tests")
+
+# Catch2 for testing
+include(FetchContent)
+
+FetchContent_Declare(
+        Catch2
+        GIT_REPOSITORY https://github.com/catchorg/Catch2.git
+        GIT_TAG        v3.5.4 # or a later release
```
Comment (alberthli) [inline tests/tests-cpp/CMakeLists.txt:9]: use spaces and not tabs (we should probably add this to the style checking for c++ code)

File: .idea/copyright/obelisk_copyright.xml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<component name="CopyrightManager">
+  <copyright>
+    <option name="notice" value="Copyright (c) &amp;#36;originalComment.match(&quot;Copyright \(c\) (\d+)&quot;, 1, &quot;-&quot;, &quot;&amp;#36;today.year&quot;)&amp;#36;today.year. Amber Lab&#10;All rights reserved." />
```
Comment (alberthli) [inline .idea/copyright/obelisk_copyright.xml:3]: nit: capitalize AMBER Lab

File: .idea/.gitignore [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
+/shelf/
+/workspace.xml
```
Comment (alberthli) [inline .idea/.gitignore:3]: add some whitespace between these blocks for readability

**Review by alberthli — CHANGES_REQUESTED** [review]
Mostly LGTM besides some minor cosmetic stuff. Also, please add a bit of documentation on the c++ side (e.g., how to build/run tests and stuff) and maybe a bit of information on the repo structure, now that it's crystallizing a bit.

File: obelisk/cpp/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    VERSION 0.0.1
+    LANGUAGES CXX
+    DESCRIPTION "Obelisk c++ utils")
+
+message(STATUS "Building obelisk_cpp")
+
+set(CMAKE_CXX_STANDARD 17 CACHE STRING "The C++ standard to use")
+
+add_subdirectory(lib1)
+add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../../tests/tests-cpp ${CMAKE_CURRENT_BINARY_DIR}/tests)
```
[the remark this answers is not in the document]
Comment (Zolkin1) [reply obelisk/cpp/CMakeLists.txt:13]: done

File: pixi.toml (hunk 2 above)
Comment (Zolkin1) [reply pixi.toml:165]: I think hyphens are more standard when interfacing with the CLI. See [here](https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html). I think the hyphens are also a bit clearer.

Lmk what you think. I can change the others to hyphens in this PR.

Comment (alberthli) [reply pixi.toml:165]: Ah I see what you mean. Yes, that makes sense. Can you do a quick scan for the underscores I have for CLI commands and turn them all into hyphens?

File: pixi.toml (hunk 3 above)
Comment (Zolkin1) [reply pixi.toml:133]: Added a build feature and another build env. I added the build feature and cuda121 and torch121, but idk if those are what you want there - lmk and I can update it.

File: tests/tests-cpp/CMakeLists.txt (hunk 6 above)
Comment (Zolkin1) [reply tests/tests-cpp/CMakeLists.txt]: this is formatting in the cmake - tbh not entirely sure how to do that right now. I can take a look though

File: docs/source/development.md [hunk 10]
```diff
@@ -40,3 +40,25 @@ pixi shell --environment dev
 The environment `dev` contains the most critical development dependencies. However, you can view the available environment sets in the `pixi.toml` to start a different environment if you would like. For example, if you are only updating the docs, you can do this by setting the environment flag to `docs`.
 
 While in the `pixi` shell and/or Docker container, all changes made in the repository or to the `~/.bashrc` file persist in your local file system. You can also use `git` with no issue to push changes.
+
+## Building and Running C++ Code
+We can easily run C++ code using `Pixi`.
+From within the `dev` enviroment, the available commands are:
```
Comment (alberthli) [inline docs/source/development.md:46]: Either combine these into one line or add whitespace between them to make it a separate paragraph.

File: docs/source/development.md [hunk 11]
```diff
@@ -40,3 +40,25 @@ pixi shell --environment dev
...
+- `cpp-ctest` will run all tests registered with `CTest`. This is the command used in the CI to verify unit tests, so every tests should be registered with CTest - see [below](#testing).
```
Comment (alberthli) [inline docs/source/development.md:47]: sp: tests -> test

File: docs/source/development.md [hunk 12]
```diff
@@ -40,3 +40,25 @@ pixi shell --environment dev
...
+- `cpp-test-dummy` will run just the dummy tests and will do so using the Catch2 framework. This command will be removed in the future and either replaced with commands to run specific tests or replaced with nothing at all.
```
Comment (alberthli) [inline docs/source/development.md:48]: annotate removal with TODO(zolkin) so we don't lose track of this.

File: docs/source/development.md [hunk 13]
```diff
@@ -40,3 +40,25 @@ pixi shell --environment dev
...
+- `cmake` which will re-build the cmake.
+- `cpp-build` which will compile the code.
+
+Note that all the commands will automatically run the commands they depend on. For this reason to re-build the cmake, compile the code, and run the tests, all we need to do is run `cpp-ctest`.
```
Comment (alberthli) [inline docs/source/development.md:52]: I think we can probably remove detail like this in the future once we explain how `pixi` works elsewhere in the docs. For now, it's probably fine, but we should have a separate PR where we streamline the docs down the road.

File: docs/source/development.md [hunk 14]
```diff
@@ -40,3 +40,25 @@ pixi shell --environment dev
...
+
+We can enter the `dev` enviroment with `pixi shell -e dev` or we can run those commands from the normal shell by pre-pending with `pixi run -e dev`. For example, to run the ctests, `pixi run -e dev cpp-ctest`.
+
+In the future we will add more commands to run other parts of the code.
+
+## C++ Code Structure
+The C++ libraries in `obelisk/` are built with CMake.
+
+All `obelisk` C++ libraries are placed in `obelisk/`. Each library should have its own folder. Within that folder there should be a `include` folder, a `CMakeLists.txt`, and the source files (i.e. not the header files - those go in `include`).
```
Comment (alberthli) [inline docs/source/development.md:61]: do you mean the C++ libraries are in `obelisk/cpp`?

File: docs/source/development.md [hunk 15]
```diff
@@ -40,3 +40,25 @@ pixi shell --environment dev
...
+
+### Testing
+Unit tests are managed with [`Catch2`](https://github.com/catchorg/Catch2). Ultimatey we plan to run the tests with [`CTest`](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Testing%20With%20CMake%20and%20CTest.html), and therefore all unit tests need to be registered with `CTest`, see [here](https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md). The tests are all placed within `tests/tests_cpp`.
```
Comment (alberthli) [inline docs/source/development.md:64]: sp: Ultimatey -> Ultimately

File: obelisk/cpp/lib1/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,7 @@
+message(STATUS "Building lib1")
+
+add_library(lib1
+        place_holder.cpp
+        include/place_holder.h)
```
Comment (alberthli) [inline obelisk/cpp/lib1/CMakeLists.txt:5]: I know you said you didn't have auto-formatting ready, but can you at least manually fix these improper tabs (or at least format it with cmake and recommit)?

File: tests/tests_cpp/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,25 @@
+message(STATUS "Building obelisk_cpp tests")
+
+# Catch2 for testing
+include(FetchContent)
+
+FetchContent_Declare(
+        Catch2
+        GIT_REPOSITORY https://github.com/catchorg/Catch2.git
+        GIT_TAG        v3.5.4 # or a later release
```
Comment (alberthli) [inline tests/tests_cpp/CMakeLists.txt:9]: same comment as above with these tabs

Comment (alberthli) [reply tests/tests_cpp/CMakeLists.txt:9]: (there are others in this file)

**Review by alberthli — APPROVED** [review]
Again, mostly LGTM with very minor comments.

**Pull request merged**